RecentActivity/src/org/sleuthkit/autopsy/recentactivity/ExtractRecycleBin.java

```diff
@@ -0,0 +1,552 @@
+                fileList = new ArrayList<>();
+                fileMap.put(fileName, fileList);
+            }
+            
+            fileList.add(rFile);
+        }
+        
+        return fileMap;
+    }
+    
+    /**
+     * Helper functions to get the user ID from the iFile parent path. User ids
+     * will be of the form S-<more characters>.
+     * 
+     * @param iFileParentPath String parent path of the iFile
+     * 
+     * @return String user id 
+     */
+    private String getUserIDFromPath(String iFileParentPath) {
+        int index = iFileParentPath.indexOf('-') - 1;
+        if (index >= 0) {
+            return (iFileParentPath.substring(index)).replace("/", "");
+        } else {
+            return "";
+        }
+    }
+    
+    /**
+     * Gets the attribute for the given type from the given artifact.
+     * 
+     * @param artifact BlackboardArtifact to get the attribute from
+     * @param type The BlackboardAttribute Type to get
+     * 
+     * @return BlackboardAttribute for given artifact and type
+     * 
+     * @throws TskCoreException 
+     */
+    private BlackboardAttribute getAttributeForArtifact(BlackboardArtifact artifact, BlackboardAttribute.ATTRIBUTE_TYPE type) throws TskCoreException{
+       return artifact.getAttribute(new BlackboardAttribute.Type(BlackboardAttribute.ATTRIBUTE_TYPE.fromID(type.getTypeID())));
+    }
+    
+    private void createRecycleBinArtifactType() throws TskCoreException {
+         try {
+            tskCase.addBlackboardArtifactType(RECYCLE_BIN_ARTIFACT_NAME, "Recycle Bin"); //NON-NLS
+        } catch (TskDataException ex) {
+            logger.log(Level.INFO, String.format("%s may have already been defined for this case", RECYCLE_BIN_ARTIFACT_NAME));
+        }
+
+    }
+    
+    /**
+     * Create the new artifact for the give rFile
+     * 
+     * @param rFile AbstractFile to create the artifact for
+     * @param type Type of artifact to create
+     * @param fileName The original path of the deleted file
+     * @param userName The name of the user that deleted the file
+     * @param dateTime The time in epoch seconds that the file was deleted
+     * 
+     * @return Newly created artifact
+     * 
+     * @throws TskCoreException 
+     */
+    private BlackboardArtifact createArtifact(AbstractFile rFile, BlackboardArtifact.Type type, String fileName, String userName, long dateTime) throws TskCoreException{
+        BlackboardArtifact bba = rFile.newArtifact(type.getTypeID());
+        bba.addAttribute(new BlackboardAttribute(TSK_PATH, getName(), fileName));
+        bba.addAttribute(new BlackboardAttribute(TSK_DATETIME_DELETED, getName(), dateTime));
+        bba.addAttribute(new BlackboardAttribute(TSK_USER_NAME, getName(), userName == null || userName.isEmpty() ? "" : userName));
+        
+        return bba;
+    }
+    
+    private AbstractFile getOrMakeFolder(SleuthkitCase skCase, FsContent dataSource, String path) throws TskCoreException {
```

Comment:
Could have FileSystem be the argument type.

Reply:
You want the function to return an FsContent?  I can do that.

Reply:
No, I meant that 'dataSource' could be of type FileSystem instead of FsContent. All we do with that argument is call getFileSystem() on it.

Reply:
SleuthkitCase.addVirtualDirectory returns a VirtualDirectory which extends SpecialDirectory which extends AbstractFile.  

The code its self doesn't need the return value from addOrMakeDirectory to be an FsContent.

Reply:
The dataSource that I pass in is the original %R file which is an AbstractFile and cannot be cast to a FileSystem.  Is there away to get a FileSystem from an AbstractFile?